<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>

Александр, доброго дня! рад приветствовать тебя) <br />
<br />
Меня зовут Николай Шавлюго. <br />И на этом этапе твоего движения к новой профессии от меня будут комментарии по написанному коду. <br />Чтобы меньше "мусорить эфир" и чтобы максимально наглядно отделяться от основного кода, есть предложение условиться в некоторых обозначениях:<br /> <br />
<u><b>ТАКОЙ ШРИФТ</b></u> - всегда начало комментария <br />
<font color='green'>такой шрифт</font> - комментарии о том, что всё ОК <br/>
<font color='orange'>такой шрифт</font> - комментарии о том, что всё ОК по результату,<br> однако есть на что обратить внимание в плане применения техник, или есть способы сделать более короткий или быстрый код <br/>
<font color='red'>такой шрифт</font> - комментарии о том, что есть критичный момент, влияющий на бизнес-результат проекта.<br/>
</font><br /><br />
<font color='green'>Моей целью является не "уличить" в не знании, а просто высказать сверху твоих знаний - свой опыт, что бы ты мог использовать его для своего дальнейшего успеха) И очень здорово будет, если тебе удастся задавать вопросы, да и вообще - всячески доставать меня, если я по каким-то причинам не приму проект:) При этом, из своего опыта скажу, хорошо и важно, когда переписка ревьюера и студента - сохраняется на следующие проекты и даже на будущую практическую деятельность.<br>
<br> Я сразу же начал общаться на ты:) но если это не очень удобно, то дай знать)) <BR> В ПУТЬ!<br>

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

**ИМПОРТЫ**

In [1]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

from sklearn.preprocessing import StandardScaler

## Загрузка и подготовка данных

In [2]:
df1 = pd.read_csv('/datasets/geo_data_0.csv')
df2 = pd.read_csv('/datasets/geo_data_1.csv')
df3 = pd.read_csv('/datasets/geo_data_2.csv')

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>ОК, данные на базе)

### Проверка данных в каждом из датасетов

**1 регион**

In [3]:
print(df1.head())
df1.info()

      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


**2 регион**

In [4]:
print(df2.head())
df2.info()

      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


**3 регион**

In [5]:
print(df3.head())
df3.info()

      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


**Вывод по блоку:**
Данные без пропусков, предобработки не требуют

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>И есть первичный аналитический осмотр). Также, можно также ещё посмотреть корреляционный анлиз, добавить графики.

### Проверка данных для построения моделей

In [6]:
numeric = ['f0', 'f1', 'f2', 'product']

**1 регион**

In [7]:
df1 = df1[numeric]

In [8]:
features1 = df1.drop(['product'], axis = 1)
target1 = df1['product']

In [9]:
features1_train, features1_valid, target1_train, target1_valid = train_test_split(features1, target1, test_size = 0.25, random_state = 12345)

In [10]:
print('Размеры тренировочной выборки:')
print(features1_train.shape)
print(target1_train.shape)

print('Размеры валидационной выборки:')
print(features1_valid.shape)
print(target1_valid.shape)


Размеры тренировочной выборки:
(75000, 3)
(75000,)
Размеры валидационной выборки:
(25000, 3)
(25000,)


In [11]:
print('Скаляризация - уравновешивание признаков...')

scaler = StandardScaler()

features1_train = scaler.fit_transform(features1_train)
features1_valid = scaler.transform(features1_valid)

Скаляризация - уравновешивание признаков...


**2 регион**

In [12]:
df2 = df2[numeric]

In [13]:
features2 = df2.drop(['product'], axis = 1)
target2 = df2['product']

In [14]:
features2_train, features2_valid, target2_train, target2_valid = train_test_split(features2, target2, test_size = 0.25, random_state = 12345)

In [15]:
print('Размеры тренировочной выборки:')
print(features2_train.shape)
print(target2_train.shape)

print('Размеры валидационной выборки:')
print(features2_valid.shape)
print(target2_valid.shape)

Размеры тренировочной выборки:
(75000, 3)
(75000,)
Размеры валидационной выборки:
(25000, 3)
(25000,)


In [16]:
print('Скаляризация - уравновешивание признаков...')

scaler = StandardScaler()

features2_train = scaler.fit_transform(features2_train)
features2_valid = scaler.transform(features2_valid)

Скаляризация - уравновешивание признаков...


**3 регион**

In [17]:
df3 = df3[numeric]

In [18]:
features3 = df3.drop(['product'], axis = 1)
target3 = df3['product']

In [19]:
features3_train, features3_valid, target3_train, target3_valid = train_test_split(features3, target3, test_size = 0.25, random_state = 12345)

In [20]:
print('Размеры тренировочной выборки:')
print(features3_train.shape)
print(target3_train.shape)

print('Размеры валидационной выборки:')
print(features3_valid.shape)
print(target3_valid.shape)

Размеры тренировочной выборки:
(75000, 3)
(75000,)
Размеры валидационной выборки:
(25000, 3)
(25000,)


In [21]:
print('Скаляризация - уравновешивание признаков...')

scaler = StandardScaler()

features3_train = scaler.fit_transform(features3_train)
features3_valid = scaler.transform(features3_valid)

Скаляризация - уравновешивание признаков...


**Вывод по блоку:**
Данные готовы для внедрения в модели!

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>Понятно. тут такой ещё момент: для всех регионов - операции оинаковые. Это значит, что однозначно можно автоматизировать расчёты. И лучше его применять)<br>
   
Можно и не делать скалярризацию - данные не сильноразличаются в порядках.... <font color='red'><br> И ещё, если делать, то во втором и третьем регионе надо проверить формулы, там описка.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='blue'> <b><u>КОММЕНТАРИЙ СТУДЕНТА</u></b>
</font>
<font color='blue'><br>Автоматизировать безусловно можно, но мне удобнее было каждый этап прописовать отдельно, повторение мать учения, хоть я и описался... Я исправил!<br>


<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font>
<font color='green'><br>это я пониманию - про повторение), но и отработать автоматизацию - тоже хорошая инвестиция времени)). ОК.

## Обучение и проверка модели

**1 регион**

In [22]:
model1 = LinearRegression()
model1.fit(features1_train, target1_train)
predictions1_train = model1.predict(features1_train)
predictions1_valid = model1.predict(features1_valid)

In [23]:
print('Средний запас сырья в обучающей выборке:', target1_train.mean())
print('Средний запас сырья в валидационной выборке:', target1_valid.mean())

Средний запас сырья в обучающей выборке: 92.64046775305694
Средний запас сырья в валидационной выборке: 92.07859674082927


In [24]:
print('RMSE модели на обучающей выборке:', mean_squared_error(target1_train, predictions1_train)**0.5)
print('Коэффициент детерминации на обучающей выборке:', r2_score(target1_train, predictions1_train))

RMSE модели на обучающей выборке: 37.72999219904026
Коэффициент детерминации на обучающей выборке: 0.27423906493940775


In [25]:
print('RMSE модели на валидационной выборке:', mean_squared_error(target1_valid, predictions1_valid)**0.5)
print('Коэффициент детерминации на валидационной выборке:', r2_score(target1_valid, predictions1_valid))

RMSE модели на валидационной выборке: 37.5794217150813
Коэффициент детерминации на валидационной выборке: 0.27994321524487786


**2 регион**

In [26]:
model2 = LinearRegression()
model2.fit(features2_train, target2_train)
predictions2_train = model2.predict(features2_train)
predictions2_valid = model2.predict(features2_valid)

In [27]:
print('Средний запас сырья в обучающей выборке:', target2_train.mean())
print('Средний запас сырья в валидационной выборке:', target2_valid.mean())

Средний запас сырья в обучающей выборке: 68.85895465854666
Средний запас сырья в валидационной выборке: 68.72313602435997


In [28]:
print('RMSE модели на обучающей выборке:', mean_squared_error(target2_train, predictions2_train)**0.5)
print('Коэффициент детерминации на обучающей выборке:', r2_score(target2_train, predictions2_train))

RMSE модели на обучающей выборке: 0.8894720773917375
Коэффициент детерминации на обучающей выборке: 0.9996247760326118


In [29]:
print('RMSE модели на валидационной выборке:', mean_squared_error(target2_valid, predictions2_valid)**0.5)
print('Коэффициент детерминации на валидационной выборке:', r2_score(target2_valid, predictions2_valid))

RMSE модели на валидационной выборке: 0.8930992867756158
Коэффициент детерминации на валидационной выборке: 0.9996233978805127


**3 регион**

In [30]:
model3 = LinearRegression()
model3.fit(features3_train, target3_train)
predictions3_train = model3.predict(features3_train)
predictions3_valid = model3.predict(features3_valid)

In [31]:
print('Средний запас сырья в обучающей выборке:', target3_train.mean())
print('Средний запас сырья в валидационной выборке:', target3_valid.mean())

Средний запас сырья в обучающей выборке: 95.03858906371522
Средний запас сырья в валидационной выборке: 94.88423280885438


In [32]:
print('RMSE модели на обучающей выборке:', mean_squared_error(target3_train, predictions3_train)**0.5)
print('Коэффициент детерминации на обучающей выборке:', r2_score(target3_train, predictions3_train))

RMSE модели на обучающей выборке: 40.06425978832384
Коэффициент детерминации на обучающей выборке: 0.19661432867329998


In [33]:
print('RMSE модели на валидационной выборке:', mean_squared_error(target3_valid, predictions3_valid)**0.5)
print('Коэффициент детерминации на валидационной выборке:', r2_score(target3_valid, predictions3_valid))

RMSE модели на валидационной выборке: 40.02970873393434
Коэффициент детерминации на валидационной выборке: 0.20524758386040443


<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>ОК, с расчётами согласен) Кстати вот по второму региону видим самую маленькую ошибку rmse. Это как раз следствие высокой корреляции f2 и целевой. Ну и по автоматизации - так же её можно сделать, т.к. для всех регионов - одинаковые расчёты.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='blue'> <b><u>КОММЕНТАРИЙ СТУДЕНТА</u></b>
</font>
<font color='blue'><br>Безусловно!<br>

**ВЫВОД**

По уже доступным нам показателям, самым привлекательным кажется регион №2.

## Подготовка к расчёту прибыли

**СПРАВОЧНЫЕ ДАННЫЕ И УСЛОВИЯ**

1) При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.

2) Бюджет на разработку скважин в регионе — 10 млрд рублей.

3) При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.

4) После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.

**УМОЗАКЛЮЧЕНИЯ И ВЫВОДЫ**

- нужно, чтобы прибыль была больше бюджета, затраченного на разработку скважин в регионе

- необходимо посчитать минимально добытый объем сырья в регионе, чтобы "перекрыть" бюджет и не допустить убытков

- необходимо найти минимальное среднее количество добычи с одной скважины, чтобы получить прибыль

Расчеты

In [34]:
min_dob = 10000000 / 450
print('Минимум,который нужно добыть:', min_dob)
min_sred = min_dob / 200
print('Минимальное среднее значение с одной скважины:', min_sred)

Минимум,который нужно добыть: 22222.222222222223
Минимальное среднее значение с одной скважины: 111.11111111111111


<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='red'><br>вот тут надо посмотреть расчёты... ожидается в районе 111.... И нужно будетт сделать выводы относительно расчётов прошлого раздела.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='blue'> <b><u>КОММЕНТАРИЙ СТУДЕНТА</u></b>
</font>
<font color='blue'><br>У нас же 10 миллиардов! Гуглил, сколько тысяч в миллиарде и потерял нуль :D<br>

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font>
<font color='green'><br>когда такие деньги, то для нас лишний ноль в одну/в другую сторону - действительно не заметно:)

Функция рассчета прибыли

In [35]:
def revenue(target, predictions, count):
    pred_sorted = predictions.sort_values(ascending=False)
    selected = target[pred_sorted.index][:count]
    return (450 * selected.sum() - 10000000)

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='red'><br>
тут просто выручка, а для прибыли нужно учесть ещё затраты. Посмотри, пожалуйста.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='blue'> <b><u>КОММЕНТАРИЙ СТУДЕНТА</u></b>
</font>
<font color='blue'><br>Теперь это выручка: прибыль минус затрата на разработку месторождений!<br>

In [36]:
def risk(values):
    counter = 0
    for v in values:
        if v >= 0:
            counter += 1
    return 100 - (counter / len(values) * 100)

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='blue'> <b><u>КОММЕНТАРИЙ СТУДЕНТА</u></b>
</font>
<font color='blue'><br>Выше - функция дилетанта для расчета рисков!<br>

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font>
<font color='green'><br>Для кого-то это дилетанство, а для кого-то будет уже достижение. Но то, что можно любой процесс, любое изделие - всегда улучшить, это ты прав. Тут нет предела совершенству, мне кажется. Вот и код выше - вполне рабочий, но можно и по другому: В саму функцию вставить строку risk=(profit_values < 0).mean()* 100 И по идее - должно тоже считать)

## Расчёт прибыли и рисков 

In [37]:
# Для всех 
state = np.random.RandomState(12345)

In [38]:
target1_valid = target1_valid.reset_index(drop=True)
target2_valid = target2_valid.reset_index(drop=True)
target3_valid = target3_valid.reset_index(drop=True)

predictions1_valid = pd.Series(predictions1_valid, index = target1_valid.index)
predictions2_valid = pd.Series(predictions2_valid, index = target2_valid.index)
predictions3_valid = pd.Series(predictions3_valid, index = target3_valid.index)

**1 регион**

In [39]:
values = []
for i in range(1000):
    target1_subsample = target1_valid.sample(n = 500, replace = True, random_state = state)
    preds1_subsample = predictions1_valid[target1_subsample.index]
    
    values.append(revenue(target1_subsample, preds1_subsample, 200))

values = pd.Series(values)
lower = values.quantile(0.025)
upper = values.quantile(0.975)

mean = values.mean()
print("Средняя выручка:", mean)
print("2.5%-квантиль:", lower)
print("97.5%-квантиль:", upper)

Средняя выручка: 425938.5269105924
2.5%-квантиль: -102090.09483793723
97.5%-квантиль: 947976.3533583689


In [40]:
print("% количества случаев расчёта, имеющих отрицательное значение прибыли:", risk(values), '%')

% количества случаев расчёта, имеющих отрицательное значение прибыли: 6.0 %


<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font><br>
<font color='red'><br>
И ещё ожидается расчёт Рисков: это % количества случаев расчёта, имеющих отрицательное значение прибыли. 

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='blue'> <b><u>КОММЕНТАРИЙ СТУДЕНТА</u></b>
</font>
<font color='blue'><br>Выше<br>

**2 регион**

In [41]:
values = []
for i in range(1000):
    target2_subsample = target2_valid.sample(n = 500, replace = True, random_state = state)
    preds2_subsample = predictions2_valid[target2_subsample.index]
    
    values.append(revenue(target2_subsample, preds2_subsample, 200))

values = pd.Series(values)
lower = values.quantile(0.025)
upper = values.quantile(0.975)

mean = values.mean()
print("Средняя выручка:", mean)
print("2.5%-квантиль:", lower)
print("97.5%-квантиль:", upper)

Средняя выручка: 518259.493697325
2.5%-квантиль: 128123.23143308601
97.5%-квантиль: 953612.9820669085


In [42]:
print("% количества случаев расчёта, имеющих отрицательное значение прибыли:", risk(values), '%')

% количества случаев расчёта, имеющих отрицательное значение прибыли: 0.29999999999999716 %


**3 регион**

In [43]:
values = []
for i in range(1000):
    target3_subsample = target3_valid.sample(n = 500, replace = True, random_state = state)
    preds3_subsample = predictions3_valid[target3_subsample.index]
    
    values.append(revenue(target3_subsample, preds3_subsample, 200))

values = pd.Series(values)
lower = values.quantile(0.025)
upper = values.quantile(0.975)

mean = values.mean()
print("Средняя выручка:", mean)
print("2.5%-квантиль:", lower)
print("97.5%-квантиль:", upper)

Средняя выручка: 420194.00534405006
2.5%-квантиль: -115852.60916001163
97.5%-квантиль: 989629.9398445741


In [44]:
print("% количества случаев расчёта, имеющих отрицательное значение прибыли:", risk(values), '%')

% количества случаев расчёта, имеющих отрицательное значение прибыли: 6.200000000000003 %


**ОБЩИЙ ВЫВОД**

Для разработки предлагаю выбрать 2 регион, так как с вероятностью 97.5% выручка будет больше 10 млрд. рублей и по расчету риска вероятность получить отрицательные значения прибыли минимальна!

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
<font color='green'><br>
Александр, на самом деле последовательная работа!<br>
И аргументация, и удобный для просмотра код - многое удалось как надо<br>
Итоги в конце - понятные и чёткие. Хорошо, что довериетльные интервала взял те, которые нужны, ОК.<br>
      
Данный проект не простой, может показаться. Потому что, его цель не просто построить прогноз по модели или прогноз продаж. А научиться выбирать лучшее управленческое решение на основе метода моделирования ситуации. В этом ключе и есть смысл её рассматривать. (Кстати, одна из программ на рынке по этой теме - AnyLogic - интересный продукт. Мне приходилось как-то смотерть её.. интересно. Можешь ролик из инета глянуть по какому-нибудь примеру с её использованием. Тогда тема управленческого моделирования может по-другому лечь... Ведь бутсреп (в лоджике егое нет) - как раз один из спобов моделирования)<br>    

<font color='red'>    
Есть несколько моментом, которые надо бы проверить:<br>
1) Расчёт функций прибыли и убытков: там сейчас просто выручка.<br>
2) также, нужен расчёт рисков: это % количества случаев расчёта, имеющих отрицательное значение прибыли<br>
3) Формулы при скаляризации...<br>
4) Расчёт точки безубыточности и её сравнение с регионами...<br>

<font color='green'>    
Спасибо)<br>
Буду ждать комментариев)

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='blue'> <b><u>КОММЕНТАРИЙ СТУДЕНТА</u></b>
</font>
<font color='blue'><br>Спасибо большое за Ваши советы и замечания! Я отстал от своей когорты, хотелось бы оперативнее досдать хвосты, поэтому не стал углубляться в автоматизацию... хотя имею представление о том, как ее сделать!
    
По поводу совета посмотреть AnyLogic - записал себе, посмотрю! Классно, что человек, непосредственно имеющий опыт решения подобных кейсов проверяет мой проект!!!
    

Вроде как, все необходимое для зачета я сделал и исправил, спасибо :)<br>

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
<font color='green'><br>Привет, Александр. Да, понял про автоматизацию. У меня тоже это было не самое сильное звено:) Но теперь точно понимаю, что важная вещь! ОК.<br>
Вроде всё по проекту теперь в точку. Спасибо за отклик.Не то, чтобы много таких кейсов было именно на эту тему. Но точно были задачи поиска решения в случаях, когда есть большая неопределённость в исходных данных.  И тогда вот данные подходы и подходы, реализованные в лоджике - помогли нащупать решение...<br>
Успехов в дальнейшем обучении! И в гонке за когортой:)

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Выполнен шаг 1: данные подготовлены
- [ ]  Выполнен шаг 2: модели обучены и проверены
    - [ ]  Данные корректно разбиты на обучающую и валидационную выборки
    - [ ]  Модели обучены, предсказания сделаны
    - [ ]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [ ]  На экране напечатаны результаты
    - [ ]  Сделаны выводы
- [ ]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [ ]  Для всех ключевых значений созданы константы Python
    - [ ]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [ ]  По предыдущему пункту сделаны выводы
    - [ ]  Написана функция расчёта прибыли
- [ ]  Выполнен шаг 4: посчитаны риски и прибыль
    - [ ]  Проведена процедура *Bootstrap*
    - [ ]  Все параметры бутстрепа соответствуют условию
    - [ ]  Найдены все нужные величины
    - [ ]  Предложен регион для разработки месторождения
    - [ ]  Выбор региона обоснован